cmd/restic/cmd_copy.go

```diff
@@ -0,0 +1,202 @@
+}
+
+var copyOptions CopyOptions
+
+func init() {
+	cmdRoot.AddCommand(cmdCopy)
+
+	f := cmdCopy.Flags()
+	f.StringVarP(&copyOptions.Repo, "repo2", "", os.Getenv("RESTIC_REPOSITORY2"), "repository to backup to or restore from (default: $RESTIC_REPOSITORY2)")
+	f.StringVarP(&copyOptions.PasswordFile, "password-file2", "", os.Getenv("RESTIC_PASSWORD_FILE2"), "read the destination repository password from a file (default: $RESTIC_PASSWORD_FILE2)")
+	f.StringVarP(&copyOptions.KeyHint, "key-hint2", "", os.Getenv("RESTIC_KEY_HINT2"), "key ID of key to try decrypting the destination repository first (default: $RESTIC_KEY_HINT2)")
+	f.StringVarP(&copyOptions.PasswordCommand, "password-command2", "", os.Getenv("RESTIC_PASSWORD_COMMAND2"), "specify a shell command to obtain a password for the destination repository (default: $RESTIC_PASSWORD_COMMAND2)")
+
+	f.StringArrayVarP(&tagOptions.Hosts, "host", "H", nil, "only consider snapshots for this `host`, when no snapshot ID is given (can be specified multiple times)")
+	f.Var(&copyOptions.Tags, "tag", "only consider snapshots which include this `taglist`, when no snapshot-ID is given")
+	f.StringArrayVar(&copyOptions.Paths, "path", nil, "only consider snapshots which include this (absolute) `path`, when no snapshot-ID is given")
+}
+
+func runCopy(opts CopyOptions, gopts GlobalOptions, args []string) error {
+	if opts.Repo == "" {
+		return errors.Fatal("Please specify a destination repository location (-repo2)")
+	}
+	var err error
+	dstGopts := gopts
+	dstGopts.Repo = opts.Repo
+	dstGopts.PasswordFile = opts.PasswordFile
+	dstGopts.KeyHint = opts.KeyHint
+	dstGopts.password, err = resolvePassword(dstGopts, "RESTIC_PASSWORD2")
+	if err != nil {
+		return err
+	}
+	dstGopts.password, err = ReadPassword(dstGopts, "enter password for destination repository: ")
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(gopts.ctx)
+	defer cancel()
+
+	srcRepo, err := OpenRepository(gopts)
+	if err != nil {
+		return err
+	}
+
+	dstRepo, err := OpenRepository(dstGopts)
+	if err != nil {
+		return err
+	}
+
+	srcLock, err := lockRepo(srcRepo)
+	defer unlockRepo(srcLock)
+	if err != nil {
+		return err
+	}
+
+	dstLock, err := lockRepo(dstRepo)
+	defer unlockRepo(dstLock)
+	if err != nil {
+		return err
+	}
+
+	debug.Log("Loading source index")
+	if err := srcRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	debug.Log("Loading destination index")
+	if err := dstRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	for sn := range FindFilteredSnapshots(ctx, srcRepo, opts.Host, opts.Tags, opts.Paths, args) {
+		Verbosef("snapshot %s of %v at %s)\n", sn.ID().Str(), sn.Paths, sn.Time)
+		Verbosef("  copy started, this may take a while...\n")
+
+		if err := copySnapshot(ctx, srcRepo, dstRepo, *sn.Tree); err != nil {
+			return err
+		}
+		debug.Log("tree copied")
+
+		if err = dstRepo.Flush(ctx); err != nil {
+			return err
+		}
+		debug.Log("flushed packs")
+
+		err = dstRepo.SaveIndex(ctx)
```

Comment:
For saving an intermediate index you could follow the lines in `internal/repository/master_index.go RebuildIndex`.
Maybe start with the check after each blob write but this then could be optimized (in a future PR).

IMO it would then also help to merge #2749.

Reply:
Now that #2773 is merged, the `SaveIndex` can simply be removed as `Flush` also handles the index.